### PR TITLE
[1.21.1] ModLoading cleanup and optimisations

### PR DESCRIPTION
**Pull request by PaintNinja**
Actually rewriting a good chunk of this code would be best but that's too much work for now, so instead here's some optimisations that shouldn't cause any breaking changes and are relatively straightforward...

- Made all applicable usages of `ModLoadingStage` `Map`s use the faster `EnumMap`
    - Was already `EnumMap` in some places before, but (likely erroneously) `HashMap` in others
- Changed some lazy-loaded singletons to use the lazy class-holder idiom as shown in this [JEP draft](<https://openjdk.org/jeps/8312611>)
    - Loading from a static final field is generally a lot quicker than a mutable instance field. Even if it's minor, the `ModLoader.get()` method is called a lot in many different contexts, so the benefit should add up nonetheless
- Avoided many allocating lambdas where trivial to do so 
- Changed the licence check to not constantly create new streams of mod info just to check one thing
    - Instead it now queries an IdentityHashSet/EmptySet that's created once and reused
- More finals, static finals and sealed heirarchies
- Used more `getFirst()`/`addFirst()`/`isEmpty()`
- Optimised `VersionSupportMatrix`
- Used `flatMap` instead of `mapMulti` where appropriate
- Used more `Map.of()`/`List.of()`
- Made more private/internal methods static
- Turned some data classes into records
- General code cleanup

### CODE REVIEW

File: fmlcore/src/main/java/net/minecraftforge/fml/ModLoader.java [hunk 1]
```diff
@@ -184,21 +184,27 @@ public void gatherAndInitializeMods(final ModWorkManager.DrivenExecutor syncExec
         this.modList = modList;
         var stateList = stateManager.getStates(ModLoadingPhase.GATHER);
         var progress = StartupMessageManager.addProgressBar("Mod Gather", stateList.stream().mapToInt(mls -> mls.size().applyAsInt(modList)).sum());
-        stateList.forEach(mls->dispatchAndHandleError(mls, syncExecutor, parallelExecutor, periodicTask, progress));
+        for (IModLoadingState mls : stateList) {
```
Comment (LexManos) [inline fmlcore/src/main/java/net/minecraftforge/fml/ModLoader.java:196]: Feel free to unfold all of the lambdas. As 1) sometimes they are faster and 2) they are always easier to debug.
The line above this is a prime example as it's calculating a sum using streams and tons of indirect functions.

File: src/main/java/net/minecraftforge/registries/RegistryObject.java [hunk 2]
```diff
@@ -189,7 +189,8 @@ public void accept(Predicate<ResourceLocation> pred) {
     @Override
     public T get() {
         T ret = this.value;
-        Objects.requireNonNull(ret, () -> "Registry Object not present: " + this.name);
+        if (ret == null)
+            throw new NullPointerException("Registry Object not present: " + this.name);
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/registries/RegistryObject.java:193]: Yes, things like this are much cleaner code in my opinion.
It also gives a convenient place to stick a breakpoint if we ever need to debug this NPE.

**Comment by LexManos** [comment]
All of this looks fine to me. Haven't done testing, but if you had im fine with saying ship it.

File: fmlcore/src/main/java/net/minecraftforge/fml/ModLoader.java (hunk 1 above)
Comment (PaintNinja) [reply fmlcore/src/main/java/net/minecraftforge/fml/ModLoader.java:196]: I think you meant to refer to line 102 of the same file, where it does creates a new stream for every iteration, which I agree is a good example of Java lambdas done wrong.

I've deliberately not unfolded all streams as there are some cases where they'd be faster than traditional techniques. For example:
```java
// stream
var streamList = this.loadingExceptions.stream().map(ModLoadingException::getModInfo).toList();

// traditional
var tradList = new ArrayList<>();
for (var loadingEx : this.loadingExceptions) {
    tradList.add(loadingEx.getModInfo());
}
```
In this example, the traditional one could result in the list being resized as more are added, along with internal array copying when that happens. The stream version on the other hand makes a pre-sized array, fills it with the mod infos without any allocations (due to a straightforward non-capturing method reference) and puts that array directly in an immutable list without any copying. The traditional one could be made to mostly do that as well but it'll be more verbose and still need an array copy to store inside a List of some kind.

Tldr streams are often overused in Forge's codebase in ways that are sometimes slow, but there are specific cases where they're actually faster.

Comment (LexManos) [reply fmlcore/src/main/java/net/minecraftforge/fml/ModLoader.java:196]: I was referring to `stateList.stream().mapToInt(mls -> mls.size().applyAsInt(modList)).sum()`
And as for your example of the sizing issues, you can minimize that by `new ArrayList<>(this.loadingExceptions.size())` but yes thats always the balance between things. Lambdas are more compact, but less understandable. 

Basically what I am saying, is that we are both agreeing that streams are overused, and if you feel like it de-streamifying things is something I'm encouraging you to do. Besides that I've already said the rest of the cleanups in this PR are fine.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/ModSorter.java [hunk 3]
```diff
@@ -242,7 +242,8 @@ private static String formatDependencyError(IModInfo.ModVersion dependency, Map<
 
     private boolean modVersionNotContained(final IModInfo.ModVersion mv, final Map<String, ArtifactVersion> modVersions)
     {
-        return !(VersionSupportMatrix.testVersionSupportMatrix(mv.getVersionRange(), mv.getModId(), "mod", (modId, range) -> modVersions.containsKey(modId) &&
+        return VersionSupportMatrix.isEnabled()
+                && !(VersionSupportMatrix.testVersionSupportMatrix(mv.getVersionRange(), mv.getModId(), "mod", (modId, range) -> modVersions.containsKey(modId) &&
```
Comment (LexManos) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/ModSorter.java]: This isn't the correct change to the logic.
Namely that the standardLookup is never called if the enabled flag is false.
Honestly that 'standard lookup' needs to be moved outside the testVersionSupportMatrix method.
And having the ENABLED flag be overrideVersion.isEmpty would make the code cleaner and function better when version numbers change.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModValidator.java [hunk 4]
```diff
@@ -106,7 +106,8 @@ public ITransformationService.Resource getModResources() {
 
     private List<EarlyLoadingException.ExceptionData> validateLanguages() {
         List<EarlyLoadingException.ExceptionData> errorData = new ArrayList<>();
-        for (Iterator<ModFile> iterator = this.candidateMods.iterator(); iterator.hasNext(); ) {
+        Iterator<ModFile> iterator = this.candidateMods.iterator();
+        while (iterator.hasNext()) {
```
Comment (LexManos) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModValidator.java]: Should be literally the same compiled code. I personally like using the former pattern for itterators. Little jumbled but its nice.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/VersionSupportMatrix.java [hunk 5]
```diff
@@ -5,48 +5,55 @@
...
 import org.apache.maven.artifact.versioning.VersionRange;
 
-import java.util.ArrayList;
-import java.util.HashMap;
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.function.BiPredicate;
 
 public class VersionSupportMatrix {
```
Comment (LexManos) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/VersionSupportMatrix.java:19]: Honestly should knock this as @Internal and make it private in 1.22

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java [hunk 6]
```diff
@@ -61,10 +61,6 @@ protected IModLocator.ModFileOrException createMod(Path path, boolean ignoreUnkn
         if (sj.moduleDataProvider().findFile(MODS_TOML).isPresent()) {
             LOGGER.debug(LogMarkers.SCAN, "Found {} mod of type {}: {}", MODS_TOML, type, path);
             mod = new ModFile(sj, this, ModFileParser::modsTomlParser);
-            if (mod.getModFileInfo().getFileProperties().containsKey(ModFileInfo.NOT_A_FORGE_MOD_PROP)) {
-                LOGGER.error(LogMarkers.SCAN, "Unable to load file \"{}\" because its mods.toml is requesting an invalid javafml loaderVersion (use \"*\" if you want to allow all versions) and is missing a forge modId dependency declaration (see the sample mods.toml in the MDK).", path);
```
Comment (LexManos) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java]: What happens when someone shoves a mod that was detected by this, for an old version (1.20.1?), into the new version?

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/ModSorter.java (hunk 3 above)
Comment (PaintNinja) [reply fmlloader/src/main/java/net/minecraftforge/fml/loading/ModSorter.java]: Yeah I got a bit confused by the double negative, I've updated ModSorter again that should fix that.

Regarding the ENABLED flag, I can do that but then if we forget to remove overrides in a new MC version, they'll still apply. That can be worked around by still checking the version before adding entries or always adding entries and checking afterwards, but unfortunately it wouldn't be much simpler.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java (hunk 6 above)
Comment (PaintNinja) [reply fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java]: I don't remember the MC versions where this detection was useful, I think it was only a couple... 1.20.2 and 1.20.4 perhaps? But not 1.20.1.

- With the detection code, if someone shoves a non-Forge 1.20.2 mod on Forge 1.21.1, it would tell the user it's not a Forge mod along with the filename.
- Without the detection code, it'll say "invalid mod file" along with the filename (same as on 1.21.1 for non-Forge mods).

My thinking was if they've put a mod jar in for the wrong MC version anyway it doesn't really matter what loader it's for in that situation, so can cleanup some rarely used code and skip its overhead. Would you like me to add it back in?

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModValidator.java (hunk 4 above)
Comment (PaintNinja) [reply fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModValidator.java]: Just checked, they're the same except for `FRAME SAME` (while loop) vs `FRAME CHOP 1` (for loop) before the `RETURN`, TIL. Would you like me to change those back to the for loop pattern?

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/ModSorter.java (hunk 3 above)
Comment (LexManos) [reply fmlloader/src/main/java/net/minecraftforge/fml/loading/ModSorter.java]: The `if (MavenVersionAdapter.createFromVersionSpec("[1.19.2]").containsVersion(version)` (checking the current version before adding entries) would be the way i would go.
Tho honestly, I would skip all that crap. And just do 
```java
var mcver = FMLLoader.versionInfo().mcVersion();
if ("1.19.2".equals(mcver))
```

Bypass all the version range stuff.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModValidator.java (hunk 4 above)
Comment (LexManos) [reply fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModValidator.java]: Yes please.
And ya, the scope of the itr variable would be slightly different, One being the end of the for loop the other being the end of the function.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java (hunk 6 above)
Comment (LexManos) [reply fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java]: I thought the issue was that they changed the mods.toml format, which caused the ModContainer {or whatever it was that read things like the license and dependencies} to error out because of them making non-optional entries optional.

I'd keep it in place, if you feel like cleaning it up I dont really care.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/ModSorter.java (hunk 3 above)
Comment (PaintNinja) [reply fmlloader/src/main/java/net/minecraftforge/fml/loading/ModSorter.java]: Agreed, I've already done that in a later commit, I think you're looking at an old version of the file ;)

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/VersionSupportMatrix.java [hunk 7]
```diff
@@ -8,11 +8,13 @@
...
 import org.apache.maven.artifact.versioning.VersionRange;
+import org.jetbrains.annotations.ApiStatus;
 
 import java.util.List;
 import java.util.Map;
 import java.util.function.BiPredicate;
 
+@ApiStatus.Internal // since 1.21.1, will be made non-public in a later MC version
 public class VersionSupportMatrix {
     private static final boolean ENABLED = FMLLoader.versionInfo().mcVersion().equals("1.21.1");
```
Comment (LexManos) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/VersionSupportMatrix.java]: I'd get rid of this ENABLED flag at all, and move it to overrideVersions.isEmpty.

Comment (PaintNinja) [reply fmlloader/src/main/java/net/minecraftforge/fml/loading/VersionSupportMatrix.java]: Oh I see what you mean now, thanks.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/VersionSupportMatrix.java [hunk 8]
```diff
@@ -10,25 +10,25 @@
...
 @ApiStatus.Internal // since 1.21.1, will be made non-public in a later MC version
 public class VersionSupportMatrix {
-    private static final boolean ENABLED = FMLLoader.versionInfo().mcVersion().equals("1.21.1");
-    private static final Map<String, List<ArtifactVersion>> overrideVersions;
+    private static final Map<String, List<ArtifactVersion>> OVERRIDE_VERSIONS;
 
     static {
-        if (ENABLED) {
-            overrideVersions = Map.ofEntries(
+        if (FMLLoader.versionInfo().mcVersion().equals("1.21.1")) {
```
Comment (LexManos) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/VersionSupportMatrix.java]: Only other slight tweak is I prefer to do "String".equals(variable) as that prevents any possibility of a NPE. Not an issue just a style note.

File: src/main/java/net/minecraftforge/internal/ForgeBindings.java [hunk 9]
```diff
@@ -15,25 +15,31 @@
...
-                return ForgeI18n.parseMessage(i18nMessage, args);
-            }
-
-            @Override
-            public String stripControlCodes(final String toStrip) {
-                return ForgeI18n.stripControlCodes(toStrip);
-            }
-        };
+        final class LazyInit {
+            private static final Supplier<I18NParser> INSTANCE = () -> new I18NParser() {
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/internal/ForgeBindings.java]: I mean I kinda get the point of making it a lazy init as the inner class ha no captures, But why is this a named class inside this method? Its just ugly code IMO to have a named class inside a method. I'd say move this out to a private inner class instead.

File: src/main/java/net/minecraftforge/client/textures/TextureAtlasSpriteLoaderManager.java [hunk 10]
```diff
@@ -13,35 +13,43 @@
...
     /**
      * Finds the loader with the given name, or null if none is registered.
      */
     @Nullable
-    public static ITextureAtlasSpriteLoader get(ResourceLocation name)
-    {
-        return LOADERS.get(name);
+    public static ITextureAtlasSpriteLoader get(ResourceLocation name) {
+        return getLoaders().get(name);
     }
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/client/textures/TextureAtlasSpriteLoaderManager.java:32]: So, I am not particularly a fan of this change. Because it is a behavior change. More specifically it means that modders can force the event to be fired by calling the get function. Instead of the modder getting a NPE {which we could make a more friendly "you call this too early" message) they get a potentially partially full map that can't be updated.

This one needs more thought as its not just a generic cleanup.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModInfo.java [hunk 11]
```diff
@@ -27,262 +27,183 @@
...
-            throw new InvalidModFileException("Invalid modId found : " + this.modId, owningFile);
+        if (!VALID_MODID.matcher(modId).matches()) {
+            LOGGER.error(LogUtils.FATAL_MARKER, "Invalid modId found in file {} - {} does not match the standard: {}", owningFile.getFile().getFilePath(), modId, VALID_MODID.pattern());
+            throw new InvalidModFileException("Invalid modId found : " + modId, owningFile);
         }
 
-        this.namespace = config.<String>getConfigElement("namespace")
-                .orElse(this.modId);
+        String namespace = config.<String>getConfigElement("namespace")
+                .orElse(modId);
```
Comment (LexManos) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModInfo.java:68]: Honestly if you were going through and redoing this. I would add 1 or 2 helper functions:
`get(config, key)` and `get(config, key, default)` which would make a lot of these smaller `get(config, "namespace", modId)`

Comment (PaintNinja) [reply fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModInfo.java:68]: The usage of the method `get(config, key)` would end up looking like `ModInfo.<String>get(config, "namespace")` because generics fun.

For the `get(config, key, default)` it could work, but then you'd have inconsistent usage as only two config elements could use it in ModInfo anyway, but thanks for the suggestion.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java (hunk 6 above)
Comment (PaintNinja) [reply fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java]: Yep, it'll still fail in roughly the same way both with or without the special handling. Only difference is the special handling shows a different error message, but only for mods made for the older MC version where the issue applied.

Comment (LexManos) [reply fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java]: Can you post a example? Just for clarity. Because like I said I thought thr error was a NPE and not a "bad mod" or something.

Comment (PaintNinja) [reply fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java]: Just tried it now, I must've misremembered, it does indeed crash with an NPE rather than gracefully showing an error. I'll add the handling back in now

Comment (PaintNinja) [reply fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java]: After further investigation, the issue was actually caused by surprising behaviour with canonical constructors in records - nothing is actually set until after the constructor has finished running, so some code needed adjusting accordingly.

**Comment by PaintNinja** [comment]
https://discord.com/channels/1129059589325852724/1129095235889270844/1274121792063012884
